Compiler: avoid gcc@* formula

xu-cheng opened the pull request:
We should never use gcc@* as the default compiler.
Otherwise, the bottle will be linked to the wrong gcc libraries.

This replaces #670, #674

maxim-belkin commented on Library/Homebrew/compilers.rb:
```diff
@@ -133,6 +133,19 @@ def find_compiler
     compilers.each do |compiler|
       case compiler
       when :gnu
+        unless OS.mac?
```
Why is this not applicable to macOS?

sjackman replied:
On macOS `gcc` is always the most recent version of GCC. On Linux `gcc` is GCC version 5, but `gcc@7` is GCC version 7.

iMichka replied:
I hope we can change this one day. With Ubuntu 18.04 gcc will be 7.x anyway, so everything will change again. I would be in favour to have the gcc formula and aliases like upstream. `gcc` should just be the lost recent gcc, because the current state is ... surprising for newcomers.

xu-cheng replied:
In fact, I think it can be argued that Mac should use this logic as well. That is to check default gcc formula first.

sjackman commented:
I'd suggest something like the following (mostly untested).

The generic implementation:
```ruby
def gnu_gcc_versions
  GNU_GCC_VERSIONS
end
```

The Linux-specific implementation should go in the file `extend/os/linux/compilers.rb`.
```ruby
def gnu_gcc_versions
  v = Formulary["gcc"].version[/\d/]
  GNU_GCC_VERSIONS - [v] + [v]
rescue FormulaUnavailableError
  GNU_GCC_VERSIONS
end
```

```diff
-        GNU_GCC_VERSIONS.reverse_each do |v|
+        gnu_gcc_versions.reverse_each do |v|
```

sjackman commented:
I'm wondering the current behaviour is a bug, and whether it should be changed or left as is. The current behaviour is that the most recent version of GCC installed is used. I think that's pretty reasonable default behaviour. To change the default behaviour is simple: `export HOMEBREW_CC=gcc-5`.

I'd like to know the extent to which GCC 5 and GCC 7 are ABI compatible.

xu-cheng commented:
This is definitely a bug. The formulae should be built under the same environment regardless whether a user runs `brew install gcc@7`. After all, the reason to install `gcc@7` is mostly likely to be used by users' personal projects instead of other formulae.

It is also worthy to mention that the `gcc` formula and `gcc@*` formulae work differently after my PRs. They install libraries in different places:
* `gcc` installs libraries at `HOMEBREW_PREFIX/lib`
* `gcc@*` install libraries at `HOMEBREW_PREFIX/lib/gcc/#{version_suffix}`

>  To change the default behaviour is simple: export HOMEBREW_CC=gcc-5.

I think `brew install` should work out of box. That is it should always use the main gcc to build formulae. This is also true for `brew test-bot` that it should always use gcc 5 to build bottles.

> I'd like to know the extent to which GCC 5 and GCC 7 are ABI compatible.

From my understanding, it is not designed to work like this. That is GCC 7 should preserve backwards ABI compatibility to older version. But not the other way around, GCC 5 is never designed to maintain forwards ABI compatibility to newer version.

xu-cheng commented:
> To change the default behaviour is simple: export HOMEBREW_CC=gcc-5.

Also, this is not entirely true. Explicitly setting `HOMEBREW_CC` will bypass the compiler selection logics. For example, it will cause formula building to be failed if we support `needs :cxx17`.

sjackman commented:
> This is also true for brew test-bot that it should always use gcc 5 to build bottles.

`gcc@7` will not be installed on `test-bot` unless a formula `depends_on "gcc@7"`.

sjackman commented:
>>To change the default behaviour is simple: `export HOMEBREW_CC=gcc-5`

>Also, this is not entirely true. Explicitly setting HOMEBREW_CC will bypass the compiler selection logics. For example, it will cause formula building to be failed if we support needs :cxx17.

Good point.

sjackman commented:
>> I'd like to know the extent to which GCC 5 and GCC 7 are ABI compatible.

> From my understanding, it is not designed to work like this. That is GCC 7 should preserve backwards ABI compatibility to older version. But not the other way around, GCC 5 is never designed to maintain forwards ABI compatibility to newer version.

I'd like to know whether a C++ executable can be compiled with GCC 7 and linked against C++ libraries compiled with GCC 5 (and installed from bottles).

A more stressing test is whether a C++ executable can be compiled with GCC 5 (and installed from a bottle) and linked against C++ libraries compiled with GCC 7.

A quick litmus test I ran just now. A simple C++ "Hello, world!" compiled with GCC 7 can be run using `libstdc++.so.6` and `libgcc_s.so.1` of GCC 5. That's not a comprehensive test, but a necessary one.

xu-cheng committed:
Compiler: avoid gcc@* formula

We should never use gcc@* as the default compiler.
Otherwise, the bottle will be linked to the wrong gcc libraries.

This replaces #670, #674

xu-cheng commented:
@sjackman I have updated the PR according to your suggestion.

sjackman commented:
I'm happy with the style of this PR. I'm not yet convinced that its purpose is necessary. If `gcc@7` is installed on a user's system, perhaps it should be the default compiler, which is the current behaviour. I'd like to know to what extent GCC 5 and GCC 7 are interoperable / ABI compatible.
@iMichka Do you have an opinion?

iMichka commented:
I put
```
export HOMEBREW_CC=gcc-5
export HOMEBREW_CCX=g++-5
```

in my .zshrc. I have seen some packages fail to build with gcc7.
Not sure if it was due to a real failure to build with gcc7, or a mismatch between gcc5 bottles and gcc7 source builds. I am sorry, I did not take any notes about what packages were failing to build like this, but it happened 2-3 times in the last months on my Ubuntu 17.10.

I personally think the default compiler for Linuxbrew should be gcc (with gcc being `gcc7`). I do not know enough about ABI mismatches. Between gcc4 and gcc5 it was quite clear, because of the c++11 changes. Now, between gcc5 and gcc7, I do not know. I bet it is ABI compatible in theory, but in practice you may always stumble on some broken stuff.

xu-cheng commented:
>  If gcc@7 is installed on a user's system, perhaps it should be the default compiler, which is the current behaviour. 

I don't think it should be like this. The reason is that `gcc` is kinda a special formula for Linuxbrew users. More specifically, we expect all users to install `gcc` formula (or use system gcc). But there will be only a few users to install `gcc@7`.

For example, if we default to use `gcc@7`, it will cause break in the following situation:
1. User installs `gcc@7` for a side project or temporary use.
2. User installs other formulae. As a result, these formulae will be built by `gcc@7`.
3. User finds that there is no use for `gcc@7` any more. So `brew uninstall gcc@7` is invoked.
4. The formulae built by `gcc@7` may be broken.

> I'd like to know to what extent GCC 5 and GCC 7 are interoperable / ABI compatible.

First of all, we should not use GCC 7 to build formulae and run them with GCC 5 libraries. This is **forward compatibility** (in contrast to the backward compatibility). **And gcc is never being designed to work in this way.**

Second, there are definitely some incompatible ABI changes. For example, C++17 vs C++11, OpenMP 4.0 vs 4.5.

Third, gcc ships many libraries. In fact, the libraries shipped in GCC 7 are different from the ones in GCC 5.
```
$ find $(brew --prefix gcc@7)/lib/ -name '*.so' -exec basename {} \; | uniq | sort | wc -l
19

$ find $(brew --prefix gcc)/lib/ -name '*.so' -exec basename {} \; | uniq | sort | wc -l
18
```

xu-cheng commented:
Ping @sjackman

sjackman commented:
Merged. Thanks for your contribution to Linuxbrew, @xu-cheng !
